packages/flutter_tools/lib/src/build_system/targets/localizations.dart

```diff
@@ -0,0 +1,272 @@
+      '--preferred-supported-locales=${options.preferredSupportedLocales}',
+  ]);
+  if (result.exitCode != 0) {
+    logger.printError(result.stdout + result.stderr as String);
+    throw Exception();
+  }
+}
+
+/// A build step that runs the generate localizations script from
+/// dev/tool/localizations.
+class GenerateLocalizationsTarget extends Target {
+  const GenerateLocalizationsTarget();
+
+  @override
+  List<Target> get dependencies => <Target>[];
+
+  @override
+  List<Source> get inputs => <Source>[
+    // This is added as a convience for developing the tool.
+    const Source.pattern('{FLUTTER_ROOT}/packages/flutter_tools/lib/src/build_system/targets/localizations.dart'),
+    // TODO(jonahwilliams): once https://github.com/flutter/flutter/issues/56321 is
+    // complete, we should add the artifact as a dependency here. Since the tool runs
+    // this code from source, looking up each dependency will be cumbersome.
+  ];
+
+  @override
+  String get name => 'gen_localizations';
+
+  @override
+  List<Source> get outputs => <Source>[];
+
+  @override
+  List<String> get depfiles => <String>['gen_localizations.d'];
+
+  @override
+  bool canSkip(Environment environment) {
+    final File configFile = environment.projectDir.childFile('l10n.yaml');
+    return !configFile.existsSync();
+  }
+
+  @override
+  Future<void> build(Environment environment) async {
+    final File configFile = environment.projectDir.childFile('l10n.yaml');
+    assert(configFile.existsSync());
+
+    final LocalizationOptions options = parseLocalizationsOptions(
+      file: configFile,
+      logger: globals.logger,
+    );
+    final DepfileService depfileService = DepfileService(
+      logger: environment.logger,
+      fileSystem: environment.fileSystem,
+    );
+
+    await generateLocalizations(
+      fileSystem: environment.fileSystem,
+      flutterRoot: environment.flutterRootDir.path,
+      logger: environment.logger,
+      processManager: environment.processManager,
+      options: options,
+      projectDir: environment.projectDir,
+      dartBinaryPath: environment.artifacts
+        .getArtifactPath(Artifact.engineDartBinary),
+      dependenciesDir: environment.buildDir,
+    );
+    final Map<String, Object> dependencies = json
+      .decode(environment.buildDir.childFile(_kDependenciesFileName).readAsStringSync()) as Map<String, Object>;
+    final Depfile depfile = Depfile(
+      <File>[
+        configFile,
+        for (dynamic inputFile in dependencies['inputs'] as List<dynamic>)
+          environment.fileSystem.file(inputFile)
+      ],
+      <File>[
+        for (dynamic outputFile in dependencies['outputs'] as List<dynamic>)
+          environment.fileSystem.file(outputFile)
+      ]
+    );
+    depfileService.writeToFile(
+      depfile,
+      environment.buildDir.childFile('gen_localizations.d'),
+    );
+  }
+}
+
+/// Typed configuration from the localizations config file.
```

Comment:
These options and their help texts will somehow need to be exposed to users once they transition to using the l10n.yaml over the command line tool by default.

Created an issue https://github.com/flutter/flutter/issues/56681 to track